KAFKA-14647: Moving TopicFilter to server-common/utils

vamossagar12 opened the pull request:
Moving TopicFilter to server-common/utils

Commit:
KAFKA-14647: Moving TopicFilter to server-common/utils

vamossagar12 commented:
@fvaleri , plz review this PR.

fvaleri commented on server-common/src/main/java/org/apache/kafka/server/util/TopicFilter.java:
```diff
@@ -0,0 +1,51 @@
+        }
+    }
+
+    public abstract boolean isTopicAllowed(String topic, boolean excludeInternalTopics);
```
It's probably useful to add an overloaded method that only takes the topic name and excludes internal topics by default. This would also simplify tests.

vamossagar12 replied:
I thought about this and I feel retaining the current scheme seems to be better as the only 2 possible states of inclusion/exclusion can be control by the single flag.

fvaleri replied:
Fair enough.

fvaleri commented on server-common/src/test/java/org/apache/kafka/server/util/TopicFilterTest.java:
```diff
@@ -0,0 +1,50 @@
+
+        IncludeList topicFilter2 = new IncludeList(".+");
+        assertTrue(topicFilter2.isTopicAllowed("alltopics", true));
+        assertFalse(topicFilter2.isTopicAllowed(Topic.GROUP_METADATA_TOPIC_NAME, true));
```
What about the other internal topic? You can use the INTERNAL_TOPICS set and test all of them.

vamossagar12 replied:
`INTERNAL_TOPICS` variables is private so I didn't use it directly. But I added the test for the other internal topic.

fvaleri commented on server-common/src/main/java/org/apache/kafka/server/util/TopicFilter.java:
```diff
@@ -0,0 +1,51 @@
+import java.util.regex.Pattern;
+import java.util.regex.PatternSyntaxException;
+
+public abstract class TopicFilter {
```
We cannot replicate the sealed behavior until Java 17, but we can have a single file containing all implementations like this:

```sh
public interface TopicFilter {
    boolean isTopicAllowed(String topic, boolean excludeInternalTopics);
    
    public static class IncludeList implements TopicFilter {
      // ...
    }
}
```

vamossagar12 replied:
I had thought about this at the time of implementation. The reason I didn't do this was that the original implementation used `sealed abstract class` which probably helps in exhaustive pattern matching. Also, I am not sure how would having the classes in the same file help in this case. Maybe I am missing something?

fvaleri replied:
It helps with reducing the number of source files and clarity as you have all implementations on the same page. This is similar to what the original scala code and TopicPartitionFilter are doing.

fvaleri commented on server-common/src/main/java/org/apache/kafka/server/util/TopicFilter.java:
```diff
@@ -0,0 +1,51 @@
+        return this.regex;
+    }
+
+    public String getRegex() {
```
Why we need this method? I think we can remove it.

vamossagar12 replied:
Oh actually MirrorMaker.scala file needed access to the regex object. I thought exposing it via a getter might be a better idea instead of exposing the field itself. Can remove if you think it isn't needed.

fvaleri replied:
Ok, you could simply return a Pattern in this case.

vamossagar12 replied:
That's a good idea. Did that. Would be pushing in some time.

fvaleri commented on server-common/src/main/java/org/apache/kafka/server/util/IncludeList.java:
```diff
@@ -0,0 +1,43 @@
+import org.slf4j.LoggerFactory;
+
+public class IncludeList extends TopicFilter {
+
```
Please, remove this extra line.

fvaleri commented on server-common/src/test/java/org/apache/kafka/server/util/TopicFilterTest.java:
```diff
@@ -0,0 +1,50 @@
+import static org.junit.jupiter.api.Assertions.assertFalse;
+
+public class TopicFilterTest {
+
```
Please, remove this extra line and all the others at the start/end of classes/methods.

fvaleri reviewed (changes requested):
Hi @vamossagar12, thanks for working on this! 

I left some comments, but there are build errors.

In this PR, it is probably a good idea to also move the TopicPartitionFilter and related code contained in GetOffsetShell. They are very similar and of general use.

Commit:
Addressing review comments

vamossagar12 commented:
> TopicPartitionFilter

@fvaleri , sorry about the build failures. I hadn't run the checkstyle for core on my local before pushing. I will work on the TopicPartitionFilter class, rest of the comments have been addressed/answered. Thanks for reviewing.

Commit:
Further review comments + moving TopicPArtitionFilter to utils

Commit:
Some more changes

Commit:
Fixing checkstyle issues

vamossagar12 commented:
Thanks @fvaleri , I made the changes.

Commit:
Fixing checkstyle with server-common

fvaleri commented on checkstyle/import-control.xml:
```diff
@@ -351,6 +351,7 @@
 
     <!-- This is required to make AlterConfigPolicyTest work. -->
     <allow pkg="org.apache.kafka.server.policy" />
+    <allow pkg="org.apache.kafka.server.util" />
```
Is this related to the above comment? If not please move it above right after joptsimple.

fvaleri commented on server-common/src/main/java/org/apache/kafka/server/util/TopicPartitionFilter.java:
```diff
@@ -0,0 +1,71 @@
+    /**
+     * Used to filter topics based on a certain criteria, for example, a set of topic names or a regular expression.
+     */
+    boolean isTopicAllowed(String topic);
```
Add an empty line after this.

fvaleri commented on server-common/src/main/java/org/apache/kafka/server/util/TopicPartitionFilter.java:
```diff
@@ -0,0 +1,71 @@
+            return filters.stream().anyMatch(tp -> tp.isTopicPartitionAllowed(partition));
+        }
+    }
+
```
Remove empty line.

fvaleri commented on server-common/src/test/java/org/apache/kafka/server/util/TopicFilterTest.java:
```diff
@@ -0,0 +1,52 @@
+        assertTrue(topicFilter4.isTopicAllowed("test-good", true));
+        assertFalse(topicFilter4.isTopicAllowed("test-bad", true));
+    }
+    
```
Remove empty line.

fvaleri reviewed (commented):
Hi @vamossagar12, I left few comments, but looks mostly good. 

I guess we also need to add some unit tests for PartitionFilter and TopicPartitionFilter, like we have for TopicFilter. They are in GetOffsetShellParsingTest, which should rather test the tool interface not filters.

Commit:
Review comments addressed

vamossagar12 commented:
Thanks @fvaleri , I addressed the review comments. Regarding unit tests for PartitionFilter and TopicPartitionFilter, I checked the tests in `GetOffsetShellParsingTest` and they are pretty comprehensive. I also think the interface per se is being tested in `GetOffsetShellTest` so I think we can let the tests remain in `GetOffsetShellParsingTest`. LMKWYT.

fvaleri commented:
Hi @vamossagar12 this looks good, but I still think we should move `GetOffsetShellParsingTest` to `TopicPartitionFilterTest` removing any reference to `GetOffsetShell` (there is already a test for that) and create a new `PartitionFilterTest`, which should be similar to `TopicFilterTest`.

vamossagar12 commented:
> Hi @vamossagar12 this looks good, but I still think we should move `GetOffsetShellParsingTest` to `TopicPartitionFilterTest` removing any reference to `GetOffsetShell` (there is already a test for that) and create a new `PartitionFilterTest`, which should be similar to `TopicFilterTest`.

Thanks @fvaleri . I tried doing this today. A problem that I see is that `GetOffsetShellParsingTest` makes use of 2 methods from `GetOffsetShell` namely `createTopicPartitionFilterWithTopicAndPartitionPattern` and `createTopicPartitionFilterWithPatternList`. Those 2 methods are also eventually used in `GetOffsetShell` when creating `topicPartitionFilter`. Ideally these 2 methods can reside in `ToolsUtils` and have both tests and `GetOffsetShell` reference it. 
I see there's also [Move GetOffsetShell to tools](https://issues.apache.org/jira/browse/KAFKA-14581). Can we do the move as part of that effort or here? WDYS?

ijuma commented:
What is the justification for this? For something to be in server-common, it needs to be used by multiple modules.

fvaleri commented:
@ijuma these classes are currently used in both tools and core modules. Not all tools will be migrated in the same release and we also have the deprecated MirrorMaker 1, which we do not plan to move.

vamossagar12 commented:
Sorry I completely missed this. Thanks @fvaleri . @ijuma do you agree with the reasoning provided by Federico?

dengziming commented:
TopicFilter and TopicPartitionFilter are very customized and simple interfaces, it's unnecessary to rewrite them currently IMO, at least currently, they are only used in a few tools.

fvaleri commented:
> In that case, would it make sense to close this PR @fvaleri ?

Why? We can discuss where to place these interfaces, but I would keep the work that has been done. It's just another step towards KAFKA-14524.

ruslankrivoshein commented:
> I would keep the work that has been done

I agree. This work is already in use in another PR.

vamossagar12 commented:
> > I would keep the work that has been done
> 
> I agree. This work is already in use in another PR.

I see. In that case, I need to fix the build issue and ask folks for reviews then. Will try to do so as soon as I can. Thanks for notifying.

Commit:
Fixing checkstyle issues

vamossagar12 commented:
@ruslankrivoshein , I have fixed the checkstyle issues. Also, I believe that the other comment [here](https://github.com/apache/kafka/pull/13158#issuecomment-1422555044) has been addressed by you in the PR https://github.com/apache/kafka/pull/13562. 
Sorry for the long wait here- it just went off my radar. @fvaleri , do you think this is looking fine now?

fvaleri commented:
@vamossagar12 LGTM. Thanks.

vamossagar12 commented:
Thanks @fvaleri . Hmm I see 101 test failures. 92 existing and 9 new. Atleast the new ones look unrelated..

vamossagar12 committed:
Merge branch 'trunk' into KAFKA-14647

fvaleri commented:
Yes, once the MirrorMaker1 dependency will be gone in 4.0.0, we can move them to tools. I added a note in [KAFKA-14705](https://issues.apache.org/jira/browse/KAFKA-14705)

vamossagar12 commented:
> I took a look at the changes and they look fine. If I understand correctly we're moving these classes to server-common while we have to keep MirrorMaker. Then in Kafka 4.0 (and assuming [KAFKA-14525](https://issues.apache.org/jira/browse/KAFKA-14525) is complete) we will be able to move all these classes to tools. Is that right?

Thanks @mimaison , yes that's the idea.

vamossagar12 commented:
@mimaison can this be merged if all changes look fine?

mimaison reviewed (approved):
Thanks!
